sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -253,6 +253,41 @@ case class GpuProjectAstExec(
   override def outputBatching: CoalesceGoal = GpuExec.outputBatching(child)
 }
 
+/**
+ * Do projections in a tiered fashion, where earlier tiers contain sub-expressions that are
+ * referenced in later tiers.  Each tier adds columns to the original batch corresponding
+ * to the output of the sub-expressions.
+ */
+case class GpuTieredProject(
+    exprSets: Seq[Seq[GpuExpression]]) extends Arm with Logging {
+
+  private def projectTier(boundExprs: Seq[Seq[GpuExpression]], cb: ColumnarBatch): ColumnarBatch = {
+    boundExprs match {
+      case Nil => {
+        GpuColumnVector.incRefCounts(cb)
+        cb
+      }
+      case exprSet :: tail => {
+        withResource(new NvtxRange("project tier", NvtxColor.ORANGE)) { _ =>
+          withResource(GpuProjectExec.project(cb, exprSet)) { projectCb =>
+            if (tail.isEmpty) {
+              projectTier(tail, projectCb)
+            } else {
+              withResource(GpuColumnVector.combineColumns(cb, projectCb)) {
```

Comment:
Agreed.  Removing eclipsed columns would be a good improvement.  It's a little complicated, because the original set of expressions could still have references to a and b that are separate from instances of (a + b), like (a + c) or (a +d).  But we should be able to determine this.   Just adding the subexpression result columns was the easier first step.

Reply:
If you have the tiers before you bind the expressions, then it is not that bad.

Look at the inputs (leaf nodes) of the output tier, this is what the output of the previous tier needs to be. If there is something missing from that tier, then add it in as a separate column that just references that column.  Then go to the next tier and look again.  Repeat until all of the tiers are done.

After that you can bind each tier the regular way to the outputs of the previous tier...

Reply:
We will do this as follow on work so I am fine with putting this in as is now.